### ISSUE TITLE
Setup WAN Federation via Mesh-gateways without exposing traffic between DCs 

### ISSUE THREAD

**Issue by dnlopes** [jcolemorrison/getting-into-consul#6]
Hey @joatmon08 @jcolemorrison 

I have been an attentive listener of your Twitch series of `getting-into-consul` (which is awesome btw, great work), but I've been struggling while setting up WAN federation via mesh gateway ([part 11](https://www.youtube.com/watch?v=3Cct7LtTGnc&list=PL81sUbsFNc5b8i2g2sB_tG-PuZxEdlDpK&index=13&t=957s)).

[Looking at the code base](https://github.com/jcolemorrison/getting-into-consul/blob/part-11/vpc-security-groups.tf#L236), it seems that for the initial setup of DC federation, all nodes between DCs must be able to reach each other. This detail is also mentioned [in the official documentation](https://developer.hashicorp.com/consul/tutorials/developer-mesh/service-mesh-gateways), in particular this part:

> Verify firewall rules
> You should also make sure the following ports are open on your firewall between each datacenter:
> 
> [8300/tcp](https://developer.hashicorp.com/consul/tutorials/developer-mesh/service-mesh-gateways#8300-tcp): all Consul agents and mesh gateways should allow traffic to each other for cross-DC WAN gossip across each datacenter. After initial federation, you can restrict this rule to traffic between a datacenter's agents and mesh gateway.
> 
> [8302/tcp](https://developer.hashicorp.com/consul/tutorials/developer-mesh/service-mesh-gateways#8302-tcp): all Consul agents and mesh gateways should allow traffic to each other for intra-DC WAN gossip across each datacenter. After initial federation, you can restrict this rule to traffic between a datacenter's agents and mesh gateway.
> 
> [8302/udp](https://developer.hashicorp.com/consul/tutorials/developer-mesh/service-mesh-gateways#8302-udp): all Consul agents and mesh gateways should allow traffic to each other for intra-DC WAN gossip across each datacenter. After initial federation, you can restrict this rule to traffic between a datacenter's agents and mesh gateway.
> 
> [8443/tcp](https://developer.hashicorp.com/consul/tutorials/developer-mesh/service-mesh-gateways#8443-tcp): mesh gateways should allow traffic from Consul agents and other mesh gateways across each datacenter.
> 
> [Envoy Listener Port/tcp](https://developer.hashicorp.com/consul/tutorials/developer-mesh/service-mesh-gateways#envoy-listener-port-tcp): sidecar proxies for each application must allow traffic from mesh gateways.

From what I understood from the above, when setting up federation for the first time, all-to-all communication is required between servers in different DCs. After federation is in place, we can restrict communications so that when cross-dc communication is required, it will always flow through the mesh-gateway.

However, this seems to (partially) defeat the purpose of federating using mesh gateways. According to another [Hashicorp documentation](https://developer.hashicorp.com/consul/docs/connect/gateways/mesh-gateway/wan-federation-via-mesh-gateways), one of the advantages of using mesh-gateway for federating multiple DCs is to increase the security of our deployment, by allowing all servers to be deployed in private subnets without public IPs:

> By activating WAN federation via mesh gateways the servers can similarly use the existing mesh gateways to reach each other without themselves being directly reachable.

But if the initial federation process requires all-to-all communication, we cannot place server nodes in a private subnet (or at least, not in the initial setup). This does not play very well with tools like Terraform where we want to automate all steps in the process.

Am I missing something, and completely misunderstood federation via mesh-gateways? Or are my assumptions are correct?

Thanks!

**Comment by dnlopes** [comment]
On the other hand, after [reading here](https://developer.hashicorp.com/consul/docs/k8s/deployment-configurations/multi-cluster), it seems that servers in a secondary DC must only be able to reach the mesh-gateway of the primary, not the primary servers directly:

> There are three networking requirements:
> 1. When Consul servers in secondary datacenters first start up, they must be able to make calls directly to the primary datacenter's mesh gateways.
> 2. Once the Consul servers in secondary datacenters have made that initial call to the primary datacenter's mesh gateways, the mesh gateways in the secondary datacenter will be able to start. From this point onwards, all communication between servers will flow first to the local mesh gateways, and then to the remote mesh gateways. This means all mesh gateways across datacenters must be able to route to one another. For example, if using a load balancer service in front of each cluster's mesh gateway pods, the load balancer IP must be routable from the other mesh gateway pods. If using a public load balancer, this is guaranteed. If using a private load balancer then you'll need to make sure that its IP/DNS address is routable from your other clusters.
> 3. If ACLs are enabled, primary clusters must be able to make requests to the Kubernetes API URLs of secondary clusters.

This makes much more sense, but the first documentation I mentioned in the OP seems to contradict this last one, thus my confusion.

So, to summarize: 
1. I'm able to put all servers in a private subnet and close the inbound traffic on those nodes to only allow traffic from the **local** VPC
2. mesh-gateway nodes need to open inbound traffic from the secondary datacenter

Is this correct?

**Comment by joatmon08** [comment]
Hello @dnlopes !!

Thanks for following along in the series. You are correct for both points - this is why we ended up recording mesh federation a few times because we couldn't quite get it working.

1. When you initially set up federation, you need enable all-to-all communication.
1. Once the initial federation gets established, the secondary servers will switch over to using the primary mesh gateway for communications (you can find this in the logs).
1. You can remove the security groups that allow Consul agent-to-agent communication across DCs. However, you still need to keep a security group rule that allows traffic from the secondary DC servers to the **primary mesh gateway**.

This does mean that your secondary DC servers must be in a subnet can route primary datacenter's subnet (either through peering, public networking, or otherwise). This means two things from a Terraform configuration perspective:

1. Possible networking configuration requirement in the form of peering / transit gateway between private subnets
2. Creating a Terraform variable that toggles `bootstrap_federate=true`. This variable will enable / disable security group rules that allows initial setup of federation and then restricts to security group to allow ingress access to the primary mesh gateway.

We're rethinking mesh federation because of the complexity of this setup. Hopefully, there will be a more simplified method going forward that allows us to set up initial federation exclusively through the mesh gateways. If you have any feedback, continue reaching out to us!

**Comment by dnlopes** [comment]
Thanks for quick reply @joatmon08. Your reply is in line with my **original post**, where I _thought_ all-to-all communication was required. But after reading [here](https://developer.hashicorp.com/consul/docs/k8s/deployment-configurations/multi-cluster#network-requirements), I was actually convinced otherwise, and that I simply understood all of these incorretly, because in that document, it says:

> When Consul servers in secondary datacenters first start up, they must be able to make calls directly to the primary datacenter's mesh gateways.

So, what I thought after reading that last documentation was:
1. servers can always be deployed in private subnets without requiring all-to-all routing
2. servers in dc2 requires direct access to `dc1 mesh gateway` for the initial bootstrap
3. after initial bootstrap, servers in dc2 can dial `dc2 mesh gateway`, which in turn will dial `dc1 mesh-gateway`
4. this would mean that after the initial setup, cross-dc communication would always go through the mesh-gateway, which is consistent with a lot of architecture diagrams I have found on official documentation on this topic.

This architecture was also aligned with my expectation: my plan was to deploy everything in private subnets, and only have the mesh-gateway publicly available, which would meet the requirement **2.** above.

But from your reply, what I understand is that it's my original post that is correct, and its the assumptions of my second comment that are actually not true 😅  This would mean that the [documentation here](https://developer.hashicorp.com/consul/docs/k8s/deployment-configurations/multi-cluster#network-requirements) is missing a network requirement: 

> all-to-all communication for the initial bootstrap

Are we on the same page here? 😄

**Comment by joatmon08** [comment]
Your original post is correct, all-to-all is **required** for initial bootstrapping of mesh federation. I double-checked with the team and it turns out connection from secondary servers to primary mesh gateway applies only in the context of cluster peering, which we didn't cover at the time of the series.

It is a good catch - very confusing, especially since it's listed under WAN federation through mesh gateways. Thanks for reporting the inconsistency in the documentation, I'll pass it onto the team to clarify and review.

**Comment by dnlopes** [comment]
Thanks for the confirmation @joatmon08. It is indeed a bit unfortunate that the initial bootstrap requires that, as it complicates a lot the deployment process. Your suggestion indeed works, but feels hackish, and will even be more complex if we really want to move the servers to a private subnet, since we would need to tell the ASG to deploy in the specific subnet depending on the `bootstrap-federation` flag. Also, in our use case, we will likely create the primary and secondary at different stages, with weeks apart, so it's absolutely required to have the inbound blocked while the secondary dc is not up and running. 

If you have an issue link or similar which I could watch to follow-up on this topic I would gladly do it. 

Thanks for the quick replies and insights!
